File: gremlin-go/driver/serializer.go
/*
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
*/

package gremlingo

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"io"
)

const graphBinaryMimeType = "application/vnd.graphbinary-v4.0"

// Serializer interface for serializing requests and deserializing responses.
type Serializer interface {
	SerializeMessage(request *RequestMessage) ([]byte, error)
	DeserializeMessage(message []byte) (Response, error)
}

// GraphBinarySerializer serializes/deserializes message to/from GraphBinary.
type GraphBinarySerializer struct {
	ser *graphBinaryTypeSerializer
}

type writer func(interface{}, io.Writer, *graphBinaryTypeSerializer) error

var serializers map[dataType]writer

func init() {
	initSerializers()
}

func newGraphBinarySerializer(handler *logHandler) *GraphBinarySerializer {
	serializer := graphBinaryTypeSerializer{handler}
	return &GraphBinarySerializer{
		ser: &serializer,
	}
}

const versionByte byte = 0x84

// SerializeMessage serializes a request message into GraphBinary format.
//
// This method is part of the Serializer interface. It is no longer used by the default driver
// flow (which now serializes requests as JSON via HttpRequest.SerializeBody), but remains
// available for custom interceptors or alternative transport protocols that require
// GraphBinary-encoded requests.
//
// Parameters:
//   - request: The request to serialize (created via MakeBytecodeRequest or MakeStringRequest)
//
// Returns:
//   - serialized: The GraphBinary-encoded request bytes
//   - error: Any serialization error encountered
func (gs *GraphBinarySerializer) SerializeMessage(request *RequestMessage) ([]byte, error) {
	finalMessage, err := gs.buildMessage(request.Gremlin, request.Fields)
	if err != nil {
		return nil, err
	}
	return finalMessage, nil
}

func (gs *GraphBinarySerializer) buildMessage(gremlin string, args map[string]interface{}) ([]byte, error) {
	buffer := bytes.Buffer{}

	// Version
	buffer.WriteByte(versionByte)

	err := gs.ser.writeValue(args, &buffer, false)
	if err != nil {
		return nil, err
	}
	err = gs.ser.writeValue(gremlin, &buffer, false)
	if err != nil {
		return nil, err
	}

	return buffer.Bytes(), nil
}

// DeserializeMessage deserializes a GraphBinary-encoded response message.
//
// This method is part of the serializer interface and is used internally by the HTTP driver.
// It is also exposed publicly to enable alternative transport protocols (gRPC, HTTP/2, etc.) to
// deserialize responses received from a Gremlin server.
//
// Parameters:
//   - message: The GraphBinary-encoded response bytes
//
// Returns:
//   - response: The deserialized Response containing results and metadata
//   - error: Any deserialization error encountered
//
// Example for alternative transports:
//
//	// Receive bytes from custom transport
//	serializer := newGraphBinarySerializer(nil)
//	resp, err := serializer.(graphBinarySerializer).DeserializeMessage(responseBytes)
//	results := resp.responseResult.data
func (gs *GraphBinarySerializer) DeserializeMessage(message []byte) (Response, error) {
	var msg Response

	if message == nil || len(message) == 0 {
		gs.ser.logHandler.log(Error, nullInput)
		return msg, newError(err0405ReadValueInvalidNullInputError)
	}

	d := NewGraphBinaryDeserializer(bytes.NewReader(message))

	// Read header
	if err := d.ReadHeader(); err != nil {
		return msg, err
	}

	results := make([]interface{}, 0)
	for {
		n, err := d.ReadFullyQualified()
		if err != nil {
			return msg, err
		}
		if n == EndOfStream() {
			break
		}
		if d.IsBulked() {
			bulkObj, err := d.ReadFullyQualified()
			if err != nil {
				return msg, err
			}
			bulk, ok := bulkObj.(int64)
			if !ok {
				return msg, fmt.Errorf("expected int64 bulk count, got %T", bulkObj)
			}
			results = append(results, &Traverser{Bulk: bulk, Value: n})
		} else {
			results = append(results, n)
		}
	}

	msg.ResponseResult.Data = results

	code, statusMsg, exception, err := d.ReadStatus()
	if err != nil {
		return msg, err
	}
	msg.ResponseStatus.code = code
	msg.ResponseStatus.message = "OK"
	if statusMsg != "" {
		msg.ResponseStatus.message = statusMsg
	}
	if exception != "" {
		msg.ResponseStatus.exception = exception
	}

	return msg, nil
}

func initSerializers() {
	serializers = map[dataType]writer{
		stringType:     stringWriter,
		bigDecimalType: bigDecimalWriter,
		bigIntegerType: bigIntWriter,
		longType:       longWriter,
		intType:        intWriter,
		shortType:      shortWriter,
		byteType: func(value interface{}, w io.Writer, typeSerializer *graphBinaryTypeSerializer) error {
			return binary.Write(w, binary.BigEndian, value.(uint8))
		},
		booleanType: func(value interface{}, w io.Writer, typeSerializer *graphBinaryTypeSerializer) error {
			return binary.Write(w, binary.BigEndian, value.(bool))
		},
		uuidType: func(value interface{}, w io.Writer, typeSerializer *graphBinaryTypeSerializer) error {
			return binary.Write(w, binary.BigEndian, value)
		},
		floatType: func(value interface{}, w io.Writer, typeSerializer *graphBinaryTypeSerializer) error {
			return binary.Write(w, binary.BigEndian, value)
		},
		doubleType: func(value interface{}, w io.Writer, typeSerializer *graphBinaryTypeSerializer) error {
			return binary.Write(w, binary.BigEndian, value)
		},
		vertexType:         vertexWriter,
		edgeType:           edgeWriter,
		propertyType:       propertyWriter,
		vertexPropertyType: vertexPropertyWriter,
		pathType:           pathWriter,
		graphType:          graphWriter,
		datetimeType:       dateTimeWriter,
		durationType:       durationWriter,
		directionType:      enumWriter,
		tType:              enumWriter,
		mergeType:          enumWriter,
		mapType:            mapWriter,
		listType:           listWriter,
		setType:            setWriter,
		byteBuffer:         byteBufferWriter,
		markerType:         markerWriter,
		compositePDTType:   pdtWriter,
	}
}

File: gremlin-python/src/main/python/gremlin_python/driver/serializer.py
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

from gremlin_python.driver.connection import GremlinServerError
from gremlin_python.process.traversal import Traverser
from gremlin_python.structure.io import graphbinaryV4
from gremlin_python.structure.io.util import Marker

__author__ = 'David M. Brown (davebshow@gmail.com), Lyndon Bauto (lyndonb@bitquilltech.com)'

"""
GraphBinaryV4
"""
class GraphBinarySerializersV4(object):
    DEFAULT_READER_CLASS = graphbinaryV4.GraphBinaryReader
    DEFAULT_WRITER_CLASS = graphbinaryV4.GraphBinaryWriter
    DEFAULT_VERSION = b"application/vnd.graphbinary-v4.0"

    int_pack = graphbinaryV4.int32_pack

    def __init__(self, reader=None, writer=None, version=None):
        if not version:
            version = self.DEFAULT_VERSION
        self._version = version
        if not reader:
            reader = self.DEFAULT_READER_CLASS()
        self._graphbinary_reader = reader
        if not writer:
            writer = self.DEFAULT_WRITER_CLASS()
        self._graphbinary_writer = writer

    @property
    def version(self):
        """Read only property"""
        return self._version

    def serialize_message(self, request_message):
        message = self.build_message(request_message.fields, request_message.gremlin)
        return message

    def build_message(self, fields, gremlin):
        message = {
            'fields': fields,
            'gremlin': gremlin
        }
        return self.finalize_message(message)

    def finalize_message(self, message):
        ba = bytearray()

        ba.extend(graphbinaryV4.uint8_pack(0x84))
        fields = message["fields"]
        ba.extend(self.int_pack(len(fields)))
        for k, v in fields.items():
            self._graphbinary_writer.to_dict(k, ba)
            self._graphbinary_writer.to_dict(v, ba)

        gremlin = message['gremlin']
        # TODO: hack to remove type code from gremlin value for V4 message format, writer doesn't seem to have a way to
        #  write value directly by passing serializer types, check back when removing bytecode
        gremlin_ba = bytearray()
        self._graphbinary_writer.to_dict(gremlin, gremlin_ba)
        ba.extend(gremlin_ba[2:])

        return bytes(ba)

    def deserialize_response_stream(self, stream):
        """
        Yield each result item from a GraphBinary response stream as it is read
        off the wire, then raise ``GremlinServerError`` if the trailing status
        is not a success code. ``Connection._receive`` dispatches through this
        method so custom response serializers can replace GraphBinary parsing.
        """
        reader = self._graphbinary_reader

        # Response header: version byte + flags byte
        stream.read(1)
        flags = stream.read(1)[0]
        bulked = flags == 0x01

        while True:
            obj = reader.to_object(stream)
            if obj == Marker.end_of_stream():
                break
            if bulked:
                bulk = reader.to_object(stream)
                yield Traverser(obj, bulk)
            else:
                yield obj

        # Trailer: status code, message, exception
        status_code = graphbinaryV4.int32_unpack(stream.read(4))
        msg_is_null = stream.read(1)[0] == 0x01
        status_message = '' if msg_is_null else reader.to_object(
            stream, graphbinaryV4.DataType.string, False)
        exc_is_null = stream.read(1)[0] == 0x01
        status_exception = '' if exc_is_null else reader.to_object(
            stream, graphbinaryV4.DataType.string, False)

        if status_code not in (200, 204):
            raise GremlinServerError({
                'code': status_code,
                'message': status_message,
                'exception': status_exception
            })
